### DISCUSSION TITLE
Possible new commands

### DISCUSSION THREAD

**Discussion by arxanas** [arxanas/git-branchless#48] [Ideas]
- [x] `git top`/`git bottom` to traverse an entire series of commits. (Names need to be revised, since the ordering may be reverse of what's expected.)
  - Actually, these might be better as flags to `git next`/`git prev`.
  - Implemented in https://github.com/arxanas/git-branchless/pull/191
- [x] Interactive `git checkout`, operating on nodes on the smartlog graph. Could include fuzzy-filtering or EasyMotion-style commit selection.
  - Implemented in #138.
- [x] `git move` as a replacement for `git rebase`, to move entire subtrees. Proposal: `--dest`, `--base`, `--restack`, and `--main` flags.
- [x] `git sync XXX` to rebase in-progress work onto main branch (or `git sync --all`, or `git sync --onto`).
- [ ] `git stage` to interactively stage a commit. Should see what `tig` does for this.
- [ ] `git split` to split the current commit (or a provided commit) using repeated `git stage`s. Splitting should not check out to a different commit, since it shouldn't affect the working copy.
- [ ] `git absorb` (there's already an implementation of `git absorb` out there, but it doesn't work unless you have a branch checked out). Absorbing should not check out to a different commit, since it shouldn't affect the working copy.
- [x] `git amend` to amend the current commit contents, but not its message. May automatically restack, or have an option to automatically restack?
  - Implemented in #170 
- [x] `git reword` to amend the current commit message, but not its contents. Should be able to safely automatically restack, since it doesn't touch the working copy.
  - Extracted to https://github.com/arxanas/git-branchless/issues/179
- [ ] `git uncommit` as a replacement for `git reset --soft HEAD^`, since the latter is unergonomic, and it doesn't hide the given commit from the smartlog. It may be desirable to unstage the previously-committed changes as well; I've never personally wanted to keep them staged after a `git reset --soft HEAD^`.

**Comment by krobelus** [comment]
> git absorb (there's already an implementation of git absorb out there, but it doesn't work unless you have a branch checked out).

https://github.com/torbiak/git-autofixup/ should do the job

**Reply by arxanas** [reply]
For a `git absorb` command, there's a good discussion here on how Mercurial does it: https://github.com/martinvonz/jj/issues/170. In particular, they use a weave-based strategy instead of a hunk-based strategy (as in [`git-absorb`](https://github.com/tummychow/git-absorb)), which seems to produce better results. I don't know which strategy `git-autofixup` uses.

That being said, it would probably be easier to implement a hunk-based strategy, especially considering that `git-absorb` is already written in Rust and licensed permissively.

**Comment by epage** [comment]
I'm torn on the value of all of these functions.  On one hand, they can be convenient, without even considering the value of preserving the commit tree (a lot of people have similar aliases).  On the other hand, the larger the specialized ecosystem, the more disconnected people become from git and this starts to look like a different system than git (more like jj).

**Reply by arxanas** [reply]
Actually, all of these commands are just to get to parity with big-company Mercurial!

But personally, I think that Git has been stagnating on the user-interface front (and it was never particularly good there anyways), so major rethinkings are warranted. I don't think it's a problem to evolve Git into a source control system approaching Jujutsu.

**Reply by steveluscher** [reply]
I would love `absorb`. It’s like the last thing I miss from big-company Mercurial.

**Comment by claytonrcarter** [comment]
If we're voting for new commands, I would love to see some commands that allow more rebase-type actions directly on the command line, in-memory (if possble?) and without having to have the branch checked out. (Like how `git move` allows rebasing of branches while not checked out.) These come to mind:

- [ ] `git drop <commit>` as an in-memory equivalent for `git rebase -i; <find commit in list>; <change command from `pick` to `drop`>; <save & quit>`
- [ ] `git squash <commit>..<commit>` ... ditto, but for squashing/combining commits on the fly w/o having to drop in and out of an editor several times (first to set the plan, then to edit the commit message)
- [x] `git move --exact` as mentioned at #176

Complaining about `rebase -i` makes me feel really lazy 😄, but `git move` feels *really* frictionless, and bringing some of that joy to dropping and squashing would feel really great.

**Reply by martinvonz** [reply]
+1 to those. I've been very happy with how they work in `jj`. Even more than those commands, I've found it very useful how `jj describe` can work on any commit. I use that functionality ~daily. So I recommend making your proposed `git reword` work on an arbitrary commit. That command (unlike the ones above) is also guaranteed to not cause conflicts, so it can easily be made to restack.

**Reply by arxanas** [reply]
> I would love to see some commands that allow more rebase-type actions directly on the command line, in-memory (if possble?) and without having to have the branch checked out

The relevant issue would be https://github.com/arxanas/git-branchless/issues/177, I think. That issue has some dependencies, and the UI isn't quite clear yet.

> git drop <commit> as an in-memory equivalent for git rebase -i; <find commit in list>; <change command from picktodrop>; <save & quit>

One day, when we get Revsets™️, I think it'll be easy enough for the user to make an alias like `git move -s "$COMMIT"- -d "$COMMIT"+ && git hide "$COMMIT"`, which would mean "move the children of this commit to be based on the parent of this commit, and then hide this commit". @martinvonz I'm curious if you can do that today in jj?

> but for squashing/combining commits on the fly w/o having to drop in and out of an editor several times (first to set the plan, then to edit the commit message)

In this case, wouldn't it suffice to use `fixup` instead of `squash`?
> I recommend making your proposed git reword work on an arbitrary commit

Yep, the issue https://github.com/arxanas/git-branchless/issues/179 mentions working on any commit (or multiple commits!).

**Reply by martinvonz** [reply]
> One day, when we get Revsets™️, I think it'll be easy enough for the user to make an alias like `git move -s "$COMMIT"- -d "$COMMIT"+ && git hide "$COMMIT"`, which would mean "move the children of this commit to be based on the parent of this commit, and then hide this commit". @martinvonz I'm curious if you can do that today in jj?

There's `jj rebase -r <source> -d <destination>`, which rebases the source commit to the destination, and also rebases descendants on top of the parents of the source commit. So it yanks a commit out of the stack and everything on top just stays in place and just falls down, like when you yank a tablecloth off a table :) `jj rebase -s <source>` (which rebases the commit along with its descendants) doesn't (yet) support specifying multiple source commits, so the command you suggested would work only if there's exactly one child commit.

> Yep, the issue #179 mentions working on any commit (or multiple commits!).

Ah, right, I had seen that before.

**Reply by claytonrcarter** [reply]
Ah, your comment led me to dig around a bit and realize that I can do this today by installing the revsets command and adding an alias like `drop='!f() { git move -s $(git revs "children($1)") -d "$1"^; git hide "$1"; }; f'`

Similarly, I suppose an alias for `move-exact` could be something like `move-exact='!f() { git co "$2"; git cherry-pick "$1"; git drop "$1"; git co -; } f'`. I'll try to play w/ these when this situation comes up again. Thanks!

> In this case, wouldn't it suffice to use fixup instead of squash?

Arg! I had forgotten about fixup! Yes, that certainly helps. That gets me a lot closer, though I still feel like I would like to point my git laser gun some commits and blast away w/o having to checkout the branch or even touch the editor.

> That command [`git reword`] is also guaranteed to not cause conflicts, so it can easily be made to restack.

Oh ... I wish you hadn't said that. It makes me feel like it might not be *too* hard to implement (and therefore potentially w/i my nascent Rust skillset) and like I might want to try to tackle it. Uh oh ... I won't be able to get this out of my head now! 😄

**Reply by pokey** [reply]
+1 for squash. Would be cool to be able to target a revset and squash it without needing an interactive rebase or checkout. Flag to set commit message / open commit message editor would be a plus

**Reply by pokey** [reply]
I guess for now, probably the easiest way to squash a revset would be the following (inspired by https://github.com/arxanas/git-branchless/wiki/Command:-git-reword#using-git-reword-to-simplify-squashing-of-branches):

```sh
revset="foo"
git query -r "roots($revset)" | xargs git reword "$revset - roots($revset)" --fixup
```

I guess the step of doing the final rebase is still a bit problematic, because it

- requires a checkout, and
- the commit to check out can't even be specified in terms of the original `$revset` because those commits have been rewritten

I wonder if this could be specified using the language proposed in https://github.com/arxanas/git-branchless/discussions/553#discussioncomment-3714647 🤔

**Reply by claytonrcarter** [reply]
> +1 for squash.

FYI this is in the works in #545. It should address the "final rebase requires a checkout" issue.

> `git query -r "roots($revset)" | xargs git reword "$revset - roots($revset)" --fixup`

1. Don't forget to add `roots($revset)` as the argument to `--fixup
2. This works if the commits in your `$revset` is contiguous, so that that `roots($revset)` resolves to a single commit, but if `root(...)` could return multiple commits you may end up with problems.

I have been, uh, exploring this space a lot recently as I've been writing a script to clean up all of the repetitive commits created by `git absorb`. The revset that I came up to find the "rootiest root" so to say is:
```
    ROOT_QUERY="
        intersection(
            stack(),
            sole(
                roots(
                    intersection(
                        ancestors($revset),
                        descendants($revset)
                     )
                )
            )
        )
    "
```

**Reply by pokey** [reply]
> FYI this is in the works in #545. It should address the "final rebase requires a checkout" issue.
> 
> >

awesome!!  looking forward to that one 😊

> 1. Don't forget to add `roots($revset)` as the argument to `--fixup

that's what the `xargs` is doing 😊

> This works if the commits in your $revset is contiguous

Ah right yeah hadn't thought about the fact that `roots()` could return something that is a descendent of another commit in the set of you have holes

Your query is neat, but I don't see why you intersect it with `stack()`

**Reply by claytonrcarter** [reply]
> that's what the xargs is doing

Oh, of course! 🤦 

>  I don't see why you intersect it with stack()

Just to prevent me from rewriting unintended commits by limiting the scope of `$revset` to my current working branch/stack. If `$revset` is crafted carefully enough, that wouldn't be needed.

> commit to check out can't even be specified in terms of the original $revset because those commits have been rewritten

I had just recently thinking about a `rewritten(...)` revset fn that would resolve possibly rewritten/rebased commits into their most current SHA/version/location/whatever. This would allow something like `git move -x abc123 && git reword 'rewritten(abc123)'`.

**Reply by pokey** [reply]
Yeah I like `rewritten`

Tho some way of referring to the commits themselves would be even better, particularly if we introduced https://github.com/arxanas/git-branchless/discussions/551

I think the extended revset language above could work, or maybe having the `move` command have a `--raw` flag where it just outputs the new commits

**Reply by arxanas** [reply]
Mercurial calls the relevant functions `predecessors` and `successors`, although I'm not sure if they're one step or transitive.

@claytonrcarter instead of `intersection(descendants(x), ancestors(y)) `, you can use `range` or just write `x:y`.

**Comment by bsima** [comment]
`git absorb` and `git split` would be awesome, I definitely could have used both of these multiple times in the last week or two..
